aggregator_core/src/datastore.rs

```diff
@@ -5920,6 +5920,12 @@ mod tests {
         assert_matches!(error, Error::DbState(_));
     }
 
+    #[rstest_reuse::apply(schema_versions_template)]
```

Comment:
This will run an instance of this test for every supported schema version. As a nitpick, this test could instead be a one-off (i.e. not parameterized by the `schema_versions_template`) which starts at the maximum known/available schema version then downgrades to 0 -- this would exercise all schema downgrades just once.

But even better than that might be a different approach: instead, modify the `schema_versions_template` to produce an `EphemeralDatastore` by applying all migrations, then downgrading to the desired version. This would effectively test that downgrade scripts actually produce working schemas. But it wouldn't detect if upgrades were broken -- we could test both the "upgrade to version" and "upgrade past version, then downgrade" approaches work, at the cost of running each parameterized test twice per version.

WDYT? (we could also take both testing approaches, to get "syntax" checking for unsupported schema versions, and "functional" testing for supported schema versions.)

Reply:
> This will run an instance of this test for every supported schema version. As a nitpick, this test could instead be a one-off (i.e. not parameterized by the schema_versions_template) which starts at the maximum known/available schema version then downgrades to 0 -- this would exercise all schema downgrades just once.

:+1:

> But even better than that might be a different approach: instead, modify the schema_versions_template to produce an EphemeralDatastore by applying all migrations, then downgrading to the desired version. This would effectively test that downgrade scripts actually produce working schemas. But it wouldn't detect if upgrades were broken -- we could test both the "upgrade to version" and "upgrade past version, then downgrade" approaches work, at the cost of running each parameterized test twice per version.

Yep! That's incoming in the next PR. (I can't take credit for the idea, @divergentdave had the same idea as you :D https://github.com/divviup/janus/issues/1279#issuecomment-1569374474).

> we could also take both testing approaches, to get "syntax" checking for unsupported schema versions, and "functional" testing for supported schema versions.

That was my intent--one test for basic syntax+functionality checking, then another set of cases that actually runs the downgraded schema against all tests.